x-pack/filebeat/input/http_endpoint/input.go

```diff
@@ -0,0 +1,279 @@
+	eventObject  *map[string]interface{} // Current event object
+	finalHandler http.HandlerFunc
+}
+
+// NewInput creates a new httpjson input
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (input.Input, error) {
+	// Extract and validate the input's configuration.
+	conf := defaultConfig()
+	if err := cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	// Build outlet for events.
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: inputContext.DynamicFields,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap input.Context's Done channel with a context.Context. This goroutine
+	// stops with the parent closes the Done channel.
+	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+	go func() {
+		defer cancelInputCtx()
+		select {
+		case <-inputContext.Done:
+		case <-inputCtx.Done():
+		}
+	}()
+
+	// If the input ever needs to be made restartable, then context would need
+	// to be recreated with each restart.
+	workerCtx, workerCancel := context.WithCancel(inputCtx)
+
+	in := &HttpEndpoint{
+		config:       conf,
+		log:          logp.NewLogger(inputName),
+		outlet:       out,
+		inputCtx:     inputCtx,
+		workerCtx:    workerCtx,
+		workerCancel: workerCancel,
+	}
+
+	// Create an instance of the HTTP server with the beat context
+	in.server, err = createServer(in)
+	if err != nil {
+		return nil, err
+	}
+
+	in.log.Infof("Initialized %v input on %v:%v", inputName, in.config.ListenAddress, in.config.ListenPort)
+
+	return in, nil
+}
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *HttpEndpoint) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go in.run()
+	})
+}
+
+func (in *HttpEndpoint) run() {
+	defer in.workerWg.Done()
+	defer in.log.Infof("%v worker has stopped.", inputName)
+	in.server.Start()
+
+	return
+}
+
+// Stops HTTP input and waits for it to finish
+func (in *HttpEndpoint) Stop() {
+	in.workerCancel()
+	in.workerWg.Wait()
+}
+
+// Wait is an alias for Stop.
+func (in *HttpEndpoint) Wait() {
+	in.Stop()
+}
+
+// If middleware validation successed, event is sent
+func (in *HttpEndpoint) sendEvent(w http.ResponseWriter, r *http.Request) {
+	event := in.outlet.OnEvent(beat.Event{
+		Timestamp: time.Now().UTC(),
+		Fields: common.MapStr{
+			in.config.Prefix: in.eventObject,
+		},
+	})
+	if !event {
+		in.sendResponse(w, http.StatusInternalServerError, in.createErrorMessage("Unable to send event"))
+		return
+	}
+
+	return
+}
+
+func (in *HttpEndpoint) sendResponse(w http.ResponseWriter, h uint, b string) {
+	w.WriteHeader(int(h))
+	w.Write([]byte(b))
+}
+
+// Triggers if middleware validation returns successful
+func (in *HttpEndpoint) apiResponse(w http.ResponseWriter, r *http.Request) {
+	objmap := make(map[string]string)
+	json.Unmarshal([]byte(in.config.ResponseHeader), &objmap)
```

Comment:
I think it would make more sense to me to make `ResponseHeader` something like `map[string][]string` a la https://golang.org/pkg/net/http/#Header. That way there's no need to json-encode the header values in the configuration.